builtin/logical/database/versioning.go

```diff
@@ -0,0 +1,292 @@
+	if db.database != nil {
+		return db.database.Type()
+	}
+	return db.legacyDatabase.Type()
+}
+
+func (db databaseVersionWrapper) Close() error {
+	if db.database != nil {
+		return db.database.Close()
+	}
+	return db.legacyDatabase.Close()
+}
+
+// /////////////////////////////////////////////////////////////////////////////////
+// Initialization
+// /////////////////////////////////////////////////////////////////////////////////
+
+func initDatabase(ctx context.Context, dbw databaseVersionWrapper, connDetails map[string]interface{}, verifyConnection bool) (newConfig map[string]interface{}, err error) {
+	if dbw.database != nil {
+		return initNewDatabase(ctx, dbw, connDetails, verifyConnection)
+	}
+	return initLegacyDatabase(ctx, dbw, connDetails, verifyConnection)
+}
+
+func initNewDatabase(ctx context.Context, dbw databaseVersionWrapper, connDetails map[string]interface{}, verifyConnection bool) (newConfig map[string]interface{}, err error) {
+	req := newdbplugin.InitializeRequest{
+		Config:           connDetails,
+		VerifyConnection: verifyConnection,
+	}
+	resp, err := dbw.database.Initialize(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	return resp.Config, nil
+}
+
+func initLegacyDatabase(ctx context.Context, dbw databaseVersionWrapper, connDetails map[string]interface{}, verifyConnection bool) (newConfig map[string]interface{}, err error) {
+	return dbw.legacyDatabase.Init(ctx, connDetails, verifyConnection)
+}
+
+// /////////////////////////////////////////////////////////////////////////////////
+// Password generation
+// /////////////////////////////////////////////////////////////////////////////////
+
+type passwordGenerator interface {
+	GeneratePasswordFromPolicy(ctx context.Context, policyName string) (password string, err error)
+}
+
+var (
+	defaultPasswordGenerator = random.DefaultStringGenerator
+)
+
+func generatePassword(ctx context.Context, dbw databaseVersionWrapper, generator passwordGenerator, passwordPolicy string) (password string, err error) {
+	// If using the legacy database, use GenerateCredentials instead of password policies
+	// This will keep the existing behavior even though passwords can be generated with a policy
+	if dbw.legacyDatabase != nil {
+		password, err := dbw.legacyDatabase.GenerateCredentials(ctx)
+		if err != nil {
+			return "", err
+		}
+		return password, nil
+	}
+
+	if passwordPolicy == "" {
+		return defaultPasswordGenerator.Generate(ctx, rand.Reader)
+	}
+	return generator.GeneratePasswordFromPolicy(ctx, passwordPolicy)
+}
+
+// /////////////////////////////////////////////////////////////////////////////////
+// Change new user
+// /////////////////////////////////////////////////////////////////////////////////
+
+func createUser(ctx context.Context, dbw databaseVersionWrapper, pg passwordGenerator, statements dbplugin.Statements, displayName, roleName string, expiration time.Time, passwordPolicy string) (username, password string, err error) {
```

Comment:
Can we make these funcs be part of `dbPluginInstance's` methods?

Reply:
This has been put into a separate wrapper struct in `version_wrapper.go` rather than on the `dbPluginInstance`.